src/uwtools/config.py

```diff
@@ -752,3 +753,39 @@ def create_config_obj(
                     raise ValueError(err_msg)
                 # Dump to file:
                 dump_method(path=outfile, cfg=config_obj)
+
+
+def print_config_section(config: dict, section_path: List[str], log: Logger) -> None:
+    """
+    Descends into the config via the given section keys, then prints the contents of the located
+    subtree as key=value pairs, one per line.
+    """
+    keys = []
+    for key in section_path:
+        keys.append(key)
+        current_path = " -> ".join(keys)
+        try:
+            subconfig = config[key]
+        except KeyError as exc:
+            msg = f"Bad config path: {current_path}"
+            log.error(msg)
+            raise UWConfigError(msg) from exc
+        if type(subconfig) in (dict, list):
+            config = subconfig
+        else:
+            msg = f"Value at {current_path} must be a dictionary or list"
+            log.error(msg)
+            raise UWConfigError(msg)
+    if isinstance(config, dict):
+        for key, value in config.items():
+            if type(value) not in (bool, float, int, str):
+                log.error(f"Non-scalar value {key} was provided")
+                raise UWConfigError("Section values provided must be scalar values")
+        for key, value in config.items():
+            print(f"{key}={value}")
+    else:
+        for value in config:
+            if type(value) not in (bool, float, int, str):
+                log.error(f"Non-scalar value {keys[-1]} was provided")
+                raise UWConfigError("Section values provided must be scalar values")
+        print(f"{keys[-1]}={config}")
```

Comment:
Since you mentioned the growing function length in standup, I can offer that you could reduce 10 lines with 4 by defining a helper function that does the `msg` / `log` / `raise` business and calling it 4 times, i.e.
```
some_good_name(f"Non-scalar value {key} was provided")
```
in this function would do the `log` and `raise` calls. That would also mean making the `log` and `raise` messages identical, but that might be a good thing.

Two other minor suggestions:
1. Since there's an outer loop (line 764) using variable `key`, it would be safer not to reuse/redefine that variable inside, in the nested loops. Perhaps use `k` and `v` for keys and values, given that those loops/scopes are so small? I probably missed a chance to give better advice in this regard in some previous comment...
2. If `key` maintains its identity throughout the loop, it should be possible (and IMO) better to replace `keys[-1]` with just `key`, which should have the same value throughout the loop.

Reply:
And if you feel like you could go full leave-the-campsite-better on this module, you could find other places where a new `some_good_name()` logging / exception-raising function could be used and update them to simplify and make code more consistent. You may not find any opportunities, but it might be worth a look.

Reply:
Regarding point 2, when I initially tried `key` instead of `keys[-1]`, I got linter complaints stating "using possibly undefined loop variable 'key'". Is there a way around this (outside of disabling it)?

Reply:
In re: item 2, sorry, you're right -- we're outside of the first `for` loop by then. The linter isn't lying. So both my items 1 and 2 are invalid, please disregard!

My (maybe?) last suggest would be to factor out the `if type(value) not in (bool, float, int, str)` check & error-generating code to a helper function to avoid the repetition of identical logic.

Reply:
What is the rule of thumb for creating helper functions? If the code is written twice, might as well have one?

Reply:
A matter of taste/intuition, I'm afraid.

I wouldn't factor `x = x + 1` and `y = y + 1` out as calls to `def add_one(n: int) -> int`. But in this case we've got that tuple of acceptable types that seems to me to be a concept worth naming, though we could also just factor that out as a local variable in the function like `scalar_types = (bool, float, int, str)` (or if `list` is acceptable, maybe `acceptable_types = (bool, float, int, list, str)`); but then the fact that we want to log a message (possibly/probably the identical message) and raise an exception all starts to look like a "chunkier" concept that might deserve its own identity. If we found that the tuple of types wasn't the same in both cases, I might not opt to factor the whole 3-4 lines out and pass the tuple into that function, but would probably at least call a single factored-out `msg` / `log` / `raise` function.

Well, one rule of thumb might be: "If I have to change one line of code for <reasons>, do I have to go change another identical line and keep them in sync?" If so, probably you want to DRY. So far, I think that's the case here.

Reply:
What about this way to re-org the code you've already got:
```
output_lines = []
for item in config:
    value = config[item] if isinstance(config, dict) else item
    if type(value) not in (bool, float, int, str):
        log_and_error(f"Non-scalar value {value} found at {current_path}")
    else:
        output_lines.append(f"{key}={value}") if isinstance(config, dict) else f"{keys[-1]}={config}"
print("\n".join(output_lines))
```
I didn't try it in a Python REPL and might have missed something, but is that useful?

Reply:
Yes, thank you, that is useful. I've implemented it with some added constraints for the printing.

Reply:
From looking at your implementation I think I realize now that we can't just build up `output_lines` no matter what, because if `config` is a `list` then `.append()`ing `f"{keys[-1]}={config}"` repeatedly isn't what we want. Have I got that right? Would it work if we did `output_lines = set()` instead, and did `output_lines.add()` instead of `output_lines.append()`? That should nullify the effect of adding `f"{keys[-1]}={config}"` multiple times and would squeeze some conditional logic out. (In your long-ago Slack question asking for recommendations for learning resources, I suggested the "Confident Ruby" video. What I'm suggesting here is in line with those ideas: Instead of using conditionals to check for and handle various cases, try to arrange for it to be safe to do _one thing_ in all cases. I'm not sure what I'm suggesting here will work, but that's the idea.)

We could also declutter slightly by taking advantage of the fact that `log_and_error()` does not return by changing e.g.
```
if type(subconfig) in (dict, list):
    config = subconfig
else:
    log_and_error(f"Value at {current_path} must be a dictionary or list", log)
```
to
```
if not type(subconfig) in (dict, list):
    log_and_error(f"Value at {current_path} must be a dictionary or list", log)
config = subconfig
```
We can never reach `config = subconfig` if `log_and_error()` is called.

I'm just nitpicking and sharing ideas now, since we're waiting for @christinaholtNOAA to comment in the discussion about list handling, above :)

Reply:
Another technique share, whether or not you want to use it:

Python's [ternary operator](https://en.wikipedia.org/wiki/Ternary_conditional_operator) can go anywhere another expression can go, so we could also
```
print("\n".join(output_lines) if isinstance(config, dict) else f"{keys[-1]}={config}")
```
if the idea I posted above about using a `set` instead of a `list` for `output_lines` doesn't pan out.

Reply:
If you are able to / want to adopt the `set` idea, `print("\n".join(sorted(output_lines)))` might be good to ensure that the order is the same run-to-run, as `set`s are unordered.

Reply:
The issue with the original re-org you provided was that because `print("\n".join(output_lines))` was simply the last line, it printed even when `config` was a list and `output_lines` was empty, so it would print an extra `\n`.

Reply:
I don't immediately see why `output_lines` would be empty, unless the `.append()` (or `.add()` with a `set`) is being guarded with `if isinstance(config, dict)`, which I'm hoping we don't need to do if we're using a `set`. I mean `config` is a container, either a `dict` or ` list`, so the block under `for item in config` should execute at least once and -- unless we call `log_and_error()` and bail out -- `.append()` or `.add()` should be called at least once. Not so?

Reply:
Sorry about that - I didn't see the second and third response yet when I commented. That was in reference to your comment [here](https://github.com/ufs-community/workflow-tools/pull/280#discussion_r1296368363) with the re-organization when `output_lines` was still a list. 

This morning I've converted `output_lines` to a set and sorted the output. The code itself is now under 20 lines as well.

Reply:
Everything related to this conversation looks good to me now. We just need to finalize treatment of lists in the other conversation.